exit/test_exit.c

```diff
@@ -0,0 +1,513 @@
+{
+	/* Return all possible least significant byte values and check whether wait() works */
+	pid_t pid;
+	int i;
+	int val[3];
+	/*
+	 * Only the least significant 8 bits shall be available to a waiting parent process,
+	 * 3 most significant bytes should be cut off so returning val[0] should resolve to returning 0, val[1] to 1 and val[2] to 2
+	 */
+	val[0] = 0x1 << 8;
+	val[1] = (0x1 << 16) + 1;
+	val[2] = (0x1 << 24) + 2;
+
+	for (i = 0; i < 256; i++) {
+		pid = fork();
+		TEST_ASSERT_GREATER_OR_EQUAL(0, pid);
+		/* child */
+		if (pid == 0) {
+			_exit(i);
+		}
+		/* parent */
+		else {
+			int status, ret;
+
+			ret = wait(&status);
+			TEST_ASSERT_EQUAL_INT(pid, ret);
+			TEST_ASSERT_EQUAL_INT(i, WEXITSTATUS(status));
+		}
+	}
+	/* Some exceeding values */
+	for (i = 0; i < 3; i++) {
+		pid = fork();
+		if (pid < 0) {
+			TEST_ASSERT_LESS_THAN_INT(0, pid);
+		}
+		/* child */
+		if (pid == 0) {
+			_exit(val[i]);
+		}
+		/* parent */
+		else {
+			int status, ret;
+
+			ret = wait(&status);
+			TEST_ASSERT_EQUAL_INT(pid, ret);
+			TEST_ASSERT_EQUAL_INT(i, WEXITSTATUS(status));
+		}
+	}
+}
+
+
+TEST(_exit, exit_status_waitpid)
+{
+	/* Check if after _exit() waitpid() works */
+	pid_t pid;
+
+	pid = fork();
+	TEST_ASSERT_GREATER_OR_EQUAL(0, pid);
+	/* child */
+	if (pid == 0) {
+		_exit(EXIT_SUCCESS);
+	}
+	/* parent */
+	else {
+		int status, ret;
+
+		ret = waitpid(pid, &status, 0);
+		TEST_ASSERT_EQUAL_INT(pid, ret);
+		TEST_ASSERT_EQUAL_INT(EXIT_SUCCESS, WEXITSTATUS(status));
+	}
+}
+
+
+TEST(_exit, chk_if_exits)
+{
+	/* Check if process terminates after _exit call */
+	pid_t pid;
+	int status;
+
+	pid = fork();
+	TEST_ASSERT_GREATER_OR_EQUAL(0, pid);
+	/* child */
+	if (pid == 0) {
+		_exit(EXIT_SUCCESS);
+	}
+	/* parent */
+	else {
+		int ret;
+
+		wait(&status);
+		TEST_ASSERT_EQUAL_INT(EXIT_SUCCESS, WEXITSTATUS(status));
+		/* Try to kill process which exited */
+		errno = 0;
+		ret = kill(pid, SIGKILL);
+		TEST_ASSERT_EQUAL_INT(-1, ret);
+		TEST_ASSERT_EQUAL_INT(ESRCH, errno);
+	}
+}
+
+
+TEST(_exit, close_streams)
+{
+	/*
+	 * Open pipe, when child exits one side of pipe should be closed,
+	 * so that call to write should fail.
+	 */
+	pid_t pid;
+	int pipefd[2];
+	int status;
+
+	TEST_ASSERT_EQUAL_INT(0, pipe(pipefd));
+	pid = fork();
+	TEST_ASSERT_GREATER_OR_EQUAL(0, pid);
+	/* child */
+	if (pid == 0) {
+		_exit(EXIT_SUCCESS);
+	}
+	/* parent */
+	else {
+		int ret;
+
+		wait(&status);
+		TEST_ASSERT_EQUAL_INT(EXIT_SUCCESS, WEXITSTATUS(status));
+		close(pipefd[0]); /* close read pipe end */
+		errno = 0;
+		ret = write(pipefd[1], TEST_STR, sizeof(TEST_STR));
+		TEST_ASSERT_EQUAL_INT(-1, ret);
+		TEST_ASSERT_EQUAL_INT(EPIPE, errno);
+		close(pipefd[1]); /* close write pipe end */
+	}
+}
+
+
+TEST(_exit, orphaned_child)
+{
+	/* Test if parent _exit affect child process */
+	pid_t pid;
+
+	pid = fork();
+	TEST_ASSERT_GREATER_OR_EQUAL(0, pid);
+	if (pid == 0) {
+		pid = fork();
+		TEST_ASSERT_GREATER_OR_EQUAL(0, pid);
```

Comment:
Assert within the child. I know that this was proposed as a desirable way of checking `fork` status, but not when you double-`fork` in the test. Now fail of the second `fork` throws off the test framework.

An instance is also present in other tests that double-`fork`, please take care of that. Maybe consult @damianloew on that matter.